src/shim.c

```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
 		return NULL;
 	}
 
-	json_len = strlen(json);
-	*len = json_len + PROXY_CTL_HEADER_SIZE;
-	proxy_ctl_msg = calloc(*len + 1, sizeof(char));
+	msg = calloc((size_t)len, sizeof(uint8_t));
 
-	if (! proxy_ctl_msg) {
+	if (! msg) {
 		abort();
 	}
 
-	set_big_endian_32((uint8_t*)proxy_ctl_msg + PROXY_CTL_HEADER_LENGTH_OFFSET, 
-				(uint32_t)(json_len));
-	strncpy(proxy_ctl_msg + PROXY_CTL_HEADER_SIZE, json, json_len);
+	set_big_endian_16(msg, PROXY_VERSION);
+	/* Header length expressed in 32 bit words.
+	 */
+	msg[HEADER_LEN_OFFSET] = (header->header_len/32);
+	msg[RES_OFFSET] =  header->type;
+	set_big_endian_16(msg + OPCODE_OFFSET, header->opcode);
+	set_big_endian_32(msg + PAYLOAD_LEN_OFFSET, header->payload_len);
+
+	if (header->payload_len) {
+		strncpy((char *)msg + PAYLOAD_OFFSET, (const char*)payload, header->payload_len);
+	}
 
-	return proxy_ctl_msg;
- }
+	return msg;
+}
 
 /*!
- * Send "hyper" payload to cc-proxy. This will be forwarded to hyperstart.
+ * Write header and payload to the proxy socket fd.
  *
- * \param fd File descriptor to send the message to(should be proxy ctl socket fd)
- * \param Hyperstart cmd id
- * \param json Json payload
+ * \param shim \ref cc_shim
+ * \param header \ref frame_header
+ * \param payload Payload to be sent
  */
-void
-send_proxy_hyper_message(int fd, const char *hyper_cmd, const char *json) {
-	char      *proxy_payload = NULL;
-	char      *proxy_command_id = "hyper";
-	char      *proxy_ctl_msg = NULL;
-	size_t     len = 0, offset = 0;
+bool
+write_message(struct cc_shim *shim, struct frame_header *header, 
+				const uint8_t *payload)
+{
+	size_t     offset = 0;
 	ssize_t    ret;
+	ssize_t    total_size;
 
-	/* cc-proxy has the following format for "hyper" payload:
-	 * {
-	 *    "id": "hyper",
-	 *    "data": {
-	 *        "hyperName": "ping",
-	 *        "data": "json payload",
-	 *    }
-	 * }
-	*/
-
-	if ( !json || fd < 0) {
-		return;
-	}
-
-
-	ret = asprintf(&proxy_payload,
-			"{\"id\":\"%s\",\"data\":{\"hyperName\":\"%s\",\"data\":%s}}",
-			proxy_command_id, hyper_cmd, json);
+	total_size = (ssize_t)((header->header_len * 32) + header->payload_len);
 
-	if (ret == -1) {
-		abort();
+	uint8_t *msg = serialize_message(header, payload, total_size);
+	if (! msg ) {
+		return false;
 	}
 
-	proxy_ctl_msg = get_proxy_ctl_msg(proxy_payload, &len);
-	free(proxy_payload);
-
-	while (offset < len) {
-		ret = write(fd, proxy_ctl_msg + offset, len-offset);
+	while (offset < total_size) {
+		ret = write(shim->proxy_sock_fd, msg + offset, (size_t)total_size-offset);
 		if (ret == EINTR) {
 			continue;
 		}
 		if (ret <= 0 ) {
-			free(proxy_ctl_msg);
+			free(msg);
 			shim_error("Error writing to proxy: %s\n", strerror(errno));
-			return;
+			return false;
 		}
 		offset += (size_t)ret;
 	}
-	free(proxy_ctl_msg);
+	free(msg);
+	return true;
+}
+
+/*!
+ * Send message to proxy.
+ *
+ * \param shim \ref cc_shim
+ * \param header \ref frame_header
+ * \param payload Payload to be sent
+ */
```

Comment:
doxygen will complain if you run it on this code due to the missing `\return`. Same comment for various other functions below.